src/coreclr/nativeaot/System.Private.CoreLib/src/Internal/Runtime/ThreadStatics.cs

```diff
@@ -137,7 +137,15 @@ private static unsafe object AllocateThreadStaticStorageForType(TypeManagerHandl
                 gcDesc = Internal.Runtime.Augments.RuntimeAugments.TypeLoaderCallbacks.GetThreadStaticGCDescForDynamicType(typeManager, typeTlsIndex);
             }
 
-            return RuntimeImports.RhNewObject((MethodTable*)gcDesc);
+            MethodTable *pMethodTable = (MethodTable*)gcDesc;
+#if FEATURE_64BIT_ALIGNMENT
+            if (pMethodTable->RequiresAlign8)
+            {
+                return InternalCalls.RhpNewFastAlign8(pMethodTable);
+            }
+#endif
+
+            return RuntimeImports.RhNewObject(pMethodTable);
```

Comment:
Why does this work AFAICT, `RhNewObject` already [takes](https://github.com/filipnavara/runtime/blob/f52856334d4929839ef97b7c11376ac51ebc6b2e/src/coreclr/nativeaot/Runtime.Base/src/System/Runtime/RuntimeExports.cs#L41-L48) `RequiresAlign8` into account.

When I looked at the root cause of this a while back, the problem was that `RequiresAlign8` isn't set on the special `gcDesc` method tables (by the compiler, at least - I didn't check dynamic statics).

Reply:
I will double check when I get to office with my ARM device. I assumed that if the test passes it should be all right but I could have overlooked something in the process.

Reply:
The relevant place to fix would be this one:
https://github.com/dotnet/runtime/blob/ffc1174638f0dd6ebba08671e931e1fce32bf4d1/src/coreclr/tools/aot/ILCompiler.Compiler/Compiler/ILScanner.cs#L760

Reply:
@SingleAccretion Good catch!

> the problem was that RequiresAlign8 isn't set on the special gcDesc method tables

https://github.com/dotnet/runtimelab/pull/2609/ worked around this problem by allocating all thread statics with 8-byte alignment.

Reply:
I agree we should revert it... I wanted to do a clean build on my RPi to figure out if I had some stale build artifacts... but unfortunately that turned out to be more painful than I expected. The combination of build tool updates and zlib-ng seems to hit a compiler bug now and I need to figure out first how to get past it:
```
  [ 30%] Building C object _deps/fetchzlibng-build/CMakeFiles/zlib.dir/arch/arm/slide_hash_armv6.c.o
  fatal error: error in backend: Cannot select: intrinsic %llvm.arm.uqsub16
```

Reply:
Turns out, it was indeed failing even on my machine. I suspect that I accidentally made the final change with adding the `if (pMethodTable->RequiresAlign8)` in a wrong VS Code window and I changed it on my local machine instead of the remote SSH session on the Raspberry Pi. I waited for the CI to be green but I also forgot that the inner loop doesn't run the smoke tests on ARM32.

I started working on a proper fix. Some early working attempt is here:
https://github.com/filipnavara/runtime/commit/0ccb0d4d1c9bad4af729f3985f1990b13b334d18. It's not exactly pretty. I'll sleep on it for a day or two before submitting it.